### PR TITLE
Merge all corrector configs

### PR DESCRIPTION
**Pull request by irinakhismatullina**

### CODE REVIEW

**Comment by irinakhismatullina** [comment]
No, here i merged only the configs for generating TyposCorrector model (and those configs that I added are actually saved with the model, as was before). Here just all the defaults for the model creating and training are stored in one place. The changes are in no way connected to the analyzer.

File: lookout/style/typos/cmdline_tools.py [hunk 1]
```diff
@@ -60,11 +60,13 @@ def cli_get_datasets(data_path: str, config: Mapping[str, Any]) -> None:
...
 def cli_train_corrector(train: str, test: str, vocabulary_path: str,
-                        frequencies_path: str, fasttext_path: str, corrector_path: str) -> None:
+                        frequencies_path: str, fasttext_path: str, corrector_path: str,
+                        config: Mapping[str, Any]) -> None:
     """Entry point for `train_and_evaluate`."""
     train = pandas.read_csv(train, index_col=0, keep_default_na=False)
     test = pandas.read_csv(test, index_col=0, keep_default_na=False)
-    model = train_and_evaluate(train, test, vocabulary_path, frequencies_path, fasttext_path)
+    model = train_and_evaluate(train, test, vocabulary_path, frequencies_path, fasttext_path,
+                               config.get("generation", {}), config.get("ranking", {}))
```
Comment (zurk) [inline lookout/style/typos/cmdline_tools.py]: let's try to avoid as many `config.get("something", {})`  as possible. You can merge the config with a `DEFAULT_CONFIG` via `merge_dicts` function.

I think with this change many configs in `__init__` methods become non-optional which is good.

File: lookout/style/typos/config.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        "neighbors_number": 0,
+        "edit_dist_number": 20,
+        "max_corrected_length": 12,
+        "start_pool_size": 64,
+        "chunksize": 256,
+    },
+    "ranking": {
+        "train_rounds": 4000,
+        "early_stopping": 200,
+        "boost_param": {"max_depth": 6,
```
Comment (zurk) [inline lookout/style/typos/config.py]:
```python
        "boost_param": {
            "max_depth": 6,
            ...
```

File: lookout/style/typos/config.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import multiprocessing
+import pathlib
+
+
+DEFAULT_DATA_DIR = pathlib.Path(__file__).parent / "data"
+DEFAULT_CONFIG = {
```
Comment (zurk) [inline lookout/style/typos/config.py]: Let's call it `DEFAULT_PRETRAIN_CONFIG` or something to not confuse with `DEFAULT_CONFIG` which is for Analyzer in FormatAnalyzer

File: lookout/style/typos/corrector.py [hunk 4]
```diff
@@ -26,83 +26,90 @@ class TyposCorrector(Model):
...
+                                         candidates.
+                       max_distance: Maximum edit distance for symspell lookup for candidates.
+                       radius: Maximum edit distance from typo allowed for candidates.
+                       max_corrected_length: Maximum length of prefix in which symspell lookup \
+                                             for typos is conducted.
+                       start_pool_size: Length of data, starting from which multiprocessing is \
+                                        desired.
+                       chunksize: Max size of a chunk for one process during multiprocessing.
+        """
+        self.ranker.set_config(config)
```
Comment (zurk) [inline lookout/style/typos/corrector.py]:
```suggestion
        self.generator.set_config(config)
```
I assume

File: lookout/style/typos/corrector.py [hunk 5]
```diff
@@ -213,33 +203,26 @@ def suggest_by_batches(self, data: pandas.DataFrame, n_candidates: int = 3,
         :param n_candidates: Number of most probable candidates to return.
         :param return_all: False to return suggestions only for corrected tokens.
         :param batch_size: Batch size.
-        :param start_pool_size: Length of data, starting from which multiprocessing is desired.
```
Comment (zurk) [inline lookout/style/typos/corrector.py]: I hope that all parameters description are moved somewhere and not just deleted.

File: lookout/style/typos/ranking.py [hunk 6]
```diff
@@ -24,41 +25,33 @@ class CandidatesRanker(Model):
...
-        self.train_rounds = self.DEFAULT_TRAIN_ROUNDS
-        self.early_stopping = self.DEFAULT_EARLY_STOPPING
-        self.boost_param = self.DEFAULT_BOOST_PARAM
+        self.set_config(config)
         self.bst = None  # type: xgb.Booster
 
-    def construct(self, boost_param: Optional[dict] = None,
-                  train_rounds: int = DEFAULT_TRAIN_ROUNDS,
-                  early_stopping: int = DEFAULT_EARLY_STOPPING) -> None:
+    def set_config(self, config: Optional[Mapping[str, Any]] = None) -> None:
```
Comment (zurk) [inline lookout/style/typos/ranking.py]: I do not think we should allow changing the config. Let's just inline this function into __init__.
Or rename it to `_set_config`,

For all `set_config` functions.

File: lookout/style/typos/cmdline_tools.py (hunk 1 above)
Comment (irinakhismatullina) [reply lookout/style/typos/cmdline_tools.py]: It's the only place i do that, here it's unavoidable: we get the config with possible `"ranking" and "generation"` fields, but the `train_and_evaluate` function expects them separately.

File: lookout/style/typos/corrector.py (hunk 5 above)
Comment (irinakhismatullina) [reply lookout/style/typos/corrector.py]: Yes that is so

File: lookout/style/typos/cmdline_tools.py (hunk 1 above)
Comment (zurk) [reply lookout/style/typos/cmdline_tools.py]:
```python
config = merge_dicts(DEFAULT_CONFIG, config)
model = train_and_evaluate(train, test, vocabulary_path, frequencies_path, fasttext_path,
                               config["generation"], config["ranking"]))
```
Can be a solution?

File: lookout/style/typos/ranking.py (hunk 6 above)
Comment (irinakhismatullina) [reply lookout/style/typos/ranking.py]: I'm not shure this is so important. Both Ranker and Generator can be used several times with different configurations after init, and it won't break anything. In the same time loading other non-config data can time some time.
TLDR i want to use that feature while experimenting with configurations, so i think it has some value

File: lookout/style/typos/cmdline_tools.py (hunk 1 above)
Comment (irinakhismatullina) [reply lookout/style/typos/cmdline_tools.py]: No, too much excessive information will be added that way

Comment (irinakhismatullina) [reply lookout/style/typos/cmdline_tools.py]: BTW about merge_configs - what would really improve the situation is to accept also None dict and treat is as empty one.

File: lookout/style/typos/ranking.py (hunk 6 above)
Comment (zurk) [reply lookout/style/typos/ranking.py]: ok

File: lookout/style/typos/cmdline_tools.py (hunk 1 above)
Comment (irinakhismatullina) [reply lookout/style/typos/cmdline_tools.py]: The colution here would be for example to do smth like that `"corrector": {"generation": {}, "ranking"}}`(and pass this `config["corrector"]` to the function). This was actually my first idea, but i decided to change it to the present one to avoid the growth of the nesting (`DEFAULT_CORRECTOR_CONFIG["corrector"]["generation"]).

Comment (zurk) [reply lookout/style/typos/cmdline_tools.py]: ok, let's keep it

File: lookout/style/typos/generation.py [hunk 7]
```diff
@@ -52,37 +54,61 @@ def __init__(self, **kwargs):
...
+                       radius: Maximum edit distance from typo allowed for candidates.
+                       max_corrected_length: Maximum length of prefix in which symspell lookup \
+                                             for typos is conducted.
+                       start_pool_size: Length of data, starting from which multiprocessing is \
+                                        desired.
+                       chunksize: Max size of a chunk for one process during multiprocessing.
+        """
+        if config is None:
+            config = {}
+        self.config = merge_dicts(DEFAULT_CORRECTOR_CONFIG["generation"], config)
```
Comment (vmarkovtsev) [inline lookout/style/typos/generation.py]: Pro tip: you can use [munchify](https://github.com/Infinidat/munch) to work with it like `self.config.chunksize` instead of `self.config["chunk_size"]` while leaving it a `dict`.

File: lookout/style/typos/ranking.py [hunk 8]
```diff
@@ -24,41 +25,33 @@ class CandidatesRanker(Model):
...
-                           "eval_metric": ["error"]}
-
-    def __init__(self, **kwargs):
-        """Initialize a new instance of CandidatesRanker class."""
+
+    def __init__(self, config: Optional[Mapping[str, Any]] = None, **kwargs):
+        """
+        Initialize a new instance of CandidatesRanker class.
+
+        :param config: Ranking configuration, options:
```
Comment (vmarkovtsev) [inline lookout/style/typos/ranking.py]: Missing "\" in the ends of lines.

File: lookout/style/typos/generation.py [hunk 9]
```diff
@@ -52,37 +54,61 @@ def __init__(self, **kwargs):
...
         self.min_freq = min(self.frequencies.values())
 
+    def set_config(self, config: Optional[Mapping[str, Any]] = None) -> None:
+        """
+        Update candidates generation config.
+
+        :param config: Candidates generation configuration, options:
+                       neighbors_number: Number of neighbors of context and typo embeddings \
+                                         to consider as candidates.
+                       edit_dist_number: Number of the most frequent tokens among tokens on \
```
Comment (vmarkovtsev) [inline lookout/style/typos/generation.py]:
```suggestion
                       edit_dist_number: Number of the most frequent tokens among tokens at \
```

File: lookout/style/typos/corrector.py [hunk 10]
```diff
@@ -26,83 +26,90 @@ class TyposCorrector(Model):
...
+        self._log.debug("%s is initialized", repr(self.ranker))
+
+    def set_generation_config(self, config: Mapping[str, Any]) -> None:
+        """
+        Update the candidates generation config.
+
+        :param config: Candidates generation configuration, options:
+                       neighbors_number: Number of neighbors of context and typo embeddings \
+                                         to consider as candidates.
+                       edit_dist_number: Number of the most frequent tokens among tokens on \
```
Comment (vmarkovtsev) [inline lookout/style/typos/corrector.py]:
```suggestion
                       edit_dist_number: Number of the most frequent tokens among tokens at \
```

File: lookout/style/typos/corrector.py [hunk 11]
```diff
@@ -26,83 +26,90 @@ class TyposCorrector(Model):
...
-                                equal edit distance from the typo to consider as candidates.
-        :param max_distance: Maximum edit distance for symspell lookup.
-        :param radius: Maximum edit distance from typo allowed for candidates.
-        """
-        self.generator.construct(vocabulary_file, frequencies_file, embeddings_file,
-                                 neighbors_number, edit_candidates, max_distance, radius)
+        :param config: Candidates generation configuration, options:
+                       neighbors_number: Number of neighbors of context and typo embeddings \
+                                         to consider as candidates.
+                       edit_dist_number: Number of the most frequent tokens among tokens on \
```
Comment (vmarkovtsev) [inline lookout/style/typos/corrector.py]:
```suggestion
                       edit_dist_number: Number of the most frequent tokens among tokens at \
```

File: lookout/style/typos/corrector.py [hunk 12]
```diff
@@ -26,83 +26,90 @@ class TyposCorrector(Model):
...
-    def __init__(self, **kwargs):
+
+    def __init__(self, ranking_config: Optional[Mapping[str, Any]] = None, **kwargs):
         """
         Initialize a new instance of TyposCorrector class.
 
-        :param kwargs: extra keyword arguments which are consumed by Model.
+        :param ranking_config: Ranking configuration, options:
+                                train_rounds: Number of training rounds.
+                                early_stopping: Early stopping parameter.
```
Comment (vmarkovtsev) [inline lookout/style/typos/corrector.py]: Would be nice if you also document the types

File: lookout/style/typos/generation.py (hunk 7 above)
Comment (vmarkovtsev) [reply lookout/style/typos/generation.py]: Not required in this PR, of course

File: lookout/style/typos/ranking.py (hunk 8 above)
Comment (irinakhismatullina) [reply lookout/style/typos/ranking.py]: Missing what exactly? "\" will put everything in one line afaik

Comment (irinakhismatullina) [reply lookout/style/typos/ranking.py]: Oh i see:)
